src/components/List/index.tsx

```diff
@@ -0,0 +1,105 @@
+import React, { useState } from 'react';
+import classNames from 'classnames';
+import isEqual from 'lodash.isequal';
+
+import styles from '@/components/List/List.css';
+
+import ListItem from '@/components/ListItem';
+
+type ListItemCategory = 'simple' | 'icon' | 'checkbox';
+type Size = 'large' | 'medium';
+
+const largeSize = 'large';
+
+interface Props<T> {
+  getItemLabel: (item: T) => React.ReactNode;
+  getItemIcon?: (item?: T) => React.ReactNode;
+  getItemKey: (item: T) => string | number;
+  getItemName?: (item?: T) => string;
+  getItemValue: (item: T) => string | number | string[];
+  id: string;
+  isItemSelected?: (item?: T) => boolean;
+  label?: string;
+  listItemCategory: ListItemCategory;
+  onClickItem?: (item: T) => void;
+  options: [T];
+  selected?: T;
+  size: Size;
+  variablesClassName?: string;
+}
+
+const List = <T extends {}>(props: Props<T>) => {
+  const {
+    label,
+    size,
+    isItemSelected,
+    getItemLabel,
+    getItemIcon,
+    getItemKey,
+    getItemName,
+    id,
+    onClickItem,
+    options,
+    selected,
+    getItemValue,
+    variablesClassName,
+    listItemCategory
+  } = props;
+
+  const [selectedItem, setSelectedItem] = useState(selected || null);
+
+  const sizeClass = `list--${size}`;
+  const listItemSizeClass = `list-item--${size}`;
+  const checkboxCategory = 'checkbox';
+
+  const checkBoxClick = item => {
+    if (isEqual(item, selectedItem)) {
+      setSelectedItem(null);
+    } else {
+      setSelectedItem(options.find(option => getItemValue(option) === getItemValue(item)) || null);
+    }
+  };
+
+  const handleClick = item => {
+    if (listItemCategory === checkboxCategory) {
+      return onClickItem ? onClickItem(item) : checkBoxClick(item);
+    }
+    return onClickItem ? onClickItem(item) : null;
+  };
+
+  return (
+    <div className={classNames(variablesClassName)}>
+      {label && <label className={classNames(styles['list-label'])}>{label}</label>}
+      <ul id={id} className={classNames(styles.list, styles[sizeClass])}>
+        {options.map(item => (
+          <li
+            key={`${getItemKey(item)}`}
```

Comment:
I have some concerns about coupling the`getItemValue` and `getItemKey`  functions in the props, what's the reason behind passing these functions as props?

Reply:
The reason for this is that the component is more flexible, this way Devs could change it and control what they want.

Reply:
@arathjz the reason to have item value and item key is that since the data is dynamic there is no way for us to know what is the value (the prop we will display in a radio button for instance) or which attribute we should use for the React key attr